src/mbgl/mtl/texture2d.cpp

```diff
@@ -75,126 +70,104 @@ size_t Texture2D::numChannels() const noexcept {
 }
 
 void Texture2D::updateSamplerConfiguration() noexcept {
-    using namespace platform;
-    samplerStateDirty = false;
-
-    MBGL_CHECK_ERROR(glTexParameteri(GL_TEXTURE_2D,
-                                     GL_TEXTURE_MIN_FILTER,
-                                     samplerState.filter == gfx::TextureFilterType::Nearest ? GL_NEAREST : GL_LINEAR));
-    MBGL_CHECK_ERROR(glTexParameteri(GL_TEXTURE_2D,
-                                     GL_TEXTURE_MAG_FILTER,
-                                     samplerState.filter == gfx::TextureFilterType::Nearest ? GL_NEAREST : GL_LINEAR));
-    MBGL_CHECK_ERROR(glTexParameteri(GL_TEXTURE_2D,
-                                     GL_TEXTURE_WRAP_S,
-                                     samplerState.wrapU == gfx::TextureWrapType::Clamp ? GL_CLAMP_TO_EDGE : GL_REPEAT));
-    MBGL_CHECK_ERROR(glTexParameteri(GL_TEXTURE_2D,
-                                     GL_TEXTURE_WRAP_T,
-                                     samplerState.wrapV == gfx::TextureWrapType::Clamp ? GL_CLAMP_TO_EDGE : GL_REPEAT));
-
-    // Keep the resource sampler configuration in sync
-    auto& glResource = static_cast<gl::TextureResource&>(*textureResource);
-    glResource.filter = samplerState.filter;
-    glResource.wrapX = samplerState.wrapU;
-    glResource.wrapY = samplerState.wrapV;
-}
-
-void Texture2D::bind(int32_t location, int32_t textureUnit) noexcept {
-    using namespace platform;
-
-    assert(gfx::MaxActiveTextureUnits > textureUnit);
-    if (gfx::MaxActiveTextureUnits <= textureUnit) return;
-
-    // Bind to the texture unit
-    context.activeTextureUnit = static_cast<uint8_t>(textureUnit);
-    context.texture[static_cast<size_t>(textureUnit)] = getTextureID();
-    boundTextureUnit = textureUnit;
+    auto samplerDescriptor = NS::RetainPtr(MTL::SamplerDescriptor::alloc()->init());
+    samplerDescriptor->setMinFilter(samplerState.filter == gfx::TextureFilterType::Nearest
+                                        ? MTL::SamplerMinMagFilterNearest
+                                        : MTL::SamplerMinMagFilterLinear);
+    samplerDescriptor->setMagFilter(samplerState.filter == gfx::TextureFilterType::Nearest
+                                        ? MTL::SamplerMinMagFilterNearest
+                                        : MTL::SamplerMinMagFilterLinear);
+    samplerDescriptor->setSAddressMode(samplerState.wrapU == gfx::TextureWrapType::Clamp
+                                           ? MTL::SamplerAddressModeClampToEdge
+                                           : MTL::SamplerAddressModeRepeat);
+    samplerDescriptor->setTAddressMode(samplerState.wrapV == gfx::TextureWrapType::Clamp
+                                           ? MTL::SamplerAddressModeClampToEdge
+                                           : MTL::SamplerAddressModeRepeat);
+    metalSamplerState = context.createMetalSamplerState(samplerDescriptor);
```

Comment:
@sjg-wdw For each metal texture I create a metal sampler state and then I pass it to fragment shader (bind function).

Reply:
That meshes nicely with the sampler state idea added earlier.  Seems logic to me!